pkg/helm/helm.go

```diff
@@ -0,0 +1,285 @@
+limitations under the License.
+*/
+
+package helm
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	metallbv1beta1 "github.com/metallb/metallb-operator/api/v1beta1"
+	"github.com/pkg/errors"
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/chart"
+	"helm.sh/helm/v3/pkg/chart/loader"
+	"helm.sh/helm/v3/pkg/cli"
+	"helm.sh/helm/v3/pkg/cli/values"
+	"helm.sh/helm/v3/pkg/getter"
+	apiext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	helmReleaseName = "metallb"
+	bgpFrr          = "frr"
+)
+
+type Parser interface {
+	GetManifests(config *metallbv1beta1.MetalLB, client client.Client,
+		isOpenShift bool) ([]*unstructured.Unstructured, error)
+	GetChart() *chart.Chart
+	GetClient() *action.Install
+}
+
+type helm struct {
+	client      *action.Install
+	envSettings *cli.EnvSettings
+	chart       *chart.Chart
+}
+
+func NewParser(helmChartPath string, helmEnvSettings *cli.EnvSettings) (Parser, error) {
+	helm := &helm{}
+	helm.envSettings = helmEnvSettings
+	helm.client = action.NewInstall(new(action.Configuration))
+	helm.client.ReleaseName = helmReleaseName
+	helm.client.Namespace = helmEnvSettings.Namespace()
+	helm.client.DryRun = true
+	helm.client.Replace = true
+	helm.client.ClientOnly = true
+	chartPath, err := helm.client.ChartPathOptions.LocateChart(helmChartPath, helmEnvSettings)
+	if err != nil {
+		return nil, err
+	}
+	helm.chart, err = loader.Load(chartPath)
+	if err != nil {
+		return nil, err
+	}
+	return helm, nil
+}
+
+func (h *helm) GetChart() *chart.Chart {
+	return h.chart
+}
+
+func (h *helm) GetClient() *action.Install {
+	return h.client
+}
+
+func (h *helm) GetManifests(config *metallbv1beta1.MetalLB, client client.Client,
+	isOpenShift bool) ([]*unstructured.Unstructured, error) {
+	chartValueOpts := &values.Options{}
+	chartValues, err := chartValueOpts.MergeValues(getter.All(h.envSettings))
+	if err != nil {
+		return nil, err
+	}
+	err = patchPrometheusValues(chartValues, client)
+	if err != nil {
+		return nil, err
+	}
+	err = patchControllerValues(chartValues, config, isOpenShift)
+	if err != nil {
+		return nil, err
+	}
+	err = patchSpeakerValues(chartValues, config)
+	if err != nil {
+		return nil, err
+	}
+	release, err := h.client.Run(h.chart, chartValues)
+	if err != nil {
+		return nil, err
+	}
+	objs, err := parseManifest(release.Manifest)
+	if err != nil {
+		return nil, err
+	}
+	return objs, nil
+}
+
+func patchPrometheusValues(chartValues map[string]interface{}, client client.Client) error {
+	metricsPort, err := valueWithDefault("METRICS_PORT", 7472)
+	if err != nil {
+		return err
+	}
+	// We shouldn't spam the api server trying to apply PodMonitors if the resource isn't installed.
+	var enablePodMonitor bool
+	if os.Getenv("DEPLOY_PODMONITORS") == "true" && podMonitorAvailable(client) {
```

Comment:
I wonder if we could get all the env variable / "environmental information" such as podMonitor from outside and pack them in another configuration struct that we pass here.
That will make these functions pure and easier to test.